main/src/ca/uwaterloo/flix/language/phase/unification/set/SetFormula.scala

```diff
@@ -652,4 +651,235 @@ object SetFormula {
   def mkDifference(f1: SetFormula, f2: SetFormula): SetFormula =
     mkInter(f1, mkCompl(f2))
 
+  //
+  // Other Functions
+  //
+
+  /**
+    * Propagation uses the following rewrites:
+    *   - `x ∩ f == x ∩ f[x -> Univ]`
+    *   - `!x ∩ f == !x ∩ f[x -> Empty]`
+    *   - `x ∪ f == x ∪ f[x -> Empty]`
+    *   - `!x ∪ f == !x ∪ f[x -> Univ]`
+    *
+    * where `x` is [[Cst]], [[Var]], or [ElemSet]].
+    *
+    * Invariant: [[ElemSet]], [[Cst]], and [[Var]] must use disjoint integers.
+    */
+  def propagation(t: SetFormula): SetFormula = propagationHelper(t, SortedMap.empty)
+
+  /**
+    * Applies set propagation to `f` where `insts` keep track of running instantiations for
+    * [[ElemSet]], [[Cst]], and [[Var]] (using the identity mapping if not present).
+    */
+  private def propagationHelper(f: SetFormula, insts: SortedMap[Int, UnivOrEmpty]): SetFormula = f match {
+    case Univ => f
+    case Empty => f
+    case Cst(c) => insts.getOrElse(c, f)
+    case Var(x) => insts.getOrElse(x, f)
+    case e@ElemSet(_) => instElemSet(e, insts)
+    case compl@Compl(f1) =>
+      val f1Prop = propagationHelper(f1, insts)
+      // Maintain and exploit reference equality for performance.
+      if (f1Prop eq f1) compl else mkCompl(f1Prop)
+
+    case Inter(elemPos0, cstsPos0, varsPos0, elemNeg0, cstsNeg0, varsNeg0, rest0) =>
+      // Compute element sets and check for early exits.
+      val elemPos = elemPos0.map(instElemSet(_, insts)) match {
+        case Some(e) => e match {
+          case Univ => None
+          case Empty => return Empty
+          case e@ElemSet(_) => Some(e)
+          case _ => ??? // Unreachable according to instElemSet.
```

Comment:
InternalCompilerException

Reply:
Also, the comment does not have enough information for me to understand.

Reply:
instElemSet says "only returns Univ/Empty/ElemSet". should I also comment that here?

Reply:
I could also add `UnivEmptyElem` to the class hierachy

Reply:
I leave it to you.

Reply:
If an exception is to be thrown, it should be ICE.

Reply:
the traits are too messy, I'd also need EmptyOrElemSet etc.